infra/lambda_functions/websocket/disconnect/app.js

```diff
@@ -44,43 +49,113 @@ exports.handler = async (event) => {
-      .scan(openDiagramsScanParams)
-      .promise();
+    connectionsToRemove = await ddb.scan(openDiagramsScanParams).promise();
     console.log("connectionsToRemove: ", connectionsToRemove);
+  } catch (e) {
+    console.log("Error when deleting connection: ", e);
+    // return {
+    //   statusCode: 500,
+    //   body: "Failed to disconnect: " + JSON.stringify(e),
+    // };
+  }
 
-    const connectionToRemove = connectionsToRemove.Items[0];
-    if (!connectionToRemove) {
-      console.log("No connection to remove from open diagrams table");
-    } else {
-      console.log("Attempting to remove connection:", connectionToRemove);
-      const deleteParamsOpenDiagramConnection = {
-        TableName: OPEN_DIAGRAMS_TABLE_NAME,
-        Key: {
-          connectionId: connectionToRemove.connectionId,
-          diagramId: connectionToRemove.diagramId,
-        },
-      };
-      try {
-        console.log(
-          "deleteParamsConnection:",
-          deleteParamsOpenDiagramConnection
-        );
-        await ddb.delete(deleteParamsOpenDiagramConnection).promise();
-        console.log("Successfully deleted connection:", connectionToRemove);
-      } catch (e) {
-        console.log("Error when deleting connection: ", e);
-      }
-    }
+  const connectionToRemove = connectionsToRemove.Items[0];
+  if (!connectionToRemove) {
+    console.log("No connection to remove from open diagrams table");
+    // return {
+    //   statusCode: 200,
+    //   body: "Disconnected",
+    // };
+  }
+  console.log("Attempting to remove connection:", connectionToRemove);
+
+  const deleteParamsOpenDiagramConnection = {
+    TableName: OPEN_DIAGRAMS_TABLE_NAME,
+    Key: {
+      connectionId: connectionToRemove.connectionId,
+      diagramId: connectionToRemove.diagramId,
+    },
+  };
+  try {
+    console.log("deleteParamsConnection:", deleteParamsOpenDiagramConnection);
+    await ddb.delete(deleteParamsOpenDiagramConnection).promise();
+    console.log("Successfully deleted connection:", connectionToRemove);
   } catch (e) {
     console.log("Error when deleting connection: ", e);
-    return {
-      statusCode: 500,
-      body: "Failed to disconnect: " + JSON.stringify(e),
-    };
   }
 
-  return { statusCode: 200, body: "Disconnected." };
+  console.log(
+    "We are about to check if we need a new master for:",
+    connectionToRemove
+  );
+
+  if (connectionToRemove.isMaster) {
+    console.log("we need to choose a new master");
+    await chooseMaster(ddb, connectionToRemove.diagramId, apigwManagementApi);
+  } else {
+    console.log("We do not a new master");
+  }
+
+  // return { statusCode: 200, body: "Disconnected." };
 };
+
+async function chooseMaster(ddb, diagramId, apigwManagementApi) {
+  /* 
+  TODO: if the newly-chosen master can't be reached, 
+   it needs to be deleted and this function should retry
+  */
+  let usersOnDiagramResult;
+  try {
+    usersOnDiagramResult = await ddb
```

Comment:
damn... I just noticed that we could end up in an odd situation... If we fail to remove the user (which can happen), then we can end up finding the user in this query...